### PR TITLE
Add fix style pimd/langevin

### PR DESCRIPTION
**Pull request by Yi-FanLi**
**Summary**

This PR adds a new fix style pimd/langevin.
It renamed the fix style proposed in https://github.com/lammps/lammps/pull/3625

It is intended to add the following features:

stochastic thermostat
isotropic / anisotropic barostat (and triclinic in the future)
multi-processing for each bead
well-optimized inter-replica communication and synchronization scheme to guarantee speed
a set of post-processing code to calculate the quantum estimators

**Author(s)**

Yifan Li
Graduate Student
Department of Chemistry, Princeton University
yifanl0716@gmail.com

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

N/A

**Implementation Notes**

**Post Submission Checklist**

- [x] The feature or features in this pull request is complete
- [x] Licensing information is complete
- [x] Corresponding author information is complete
- [x] The source code follows the LAMMPS formatting guidelines
- [x] Suitable new documentation files and/or updates to the existing docs are included
- [x] The added/updated documentation is integrated and tested with the documentation build system
- [x] The feature has been verified to work with the conventional build system
- [x] The feature has been verified to work with the CMake based build system
- [x] One or more example input decks are included

**Further Information, Files, and Links**

TODO list for making the pull request ready for review
- [x] remove all dead code, unused keywords, unused class members, commented out code
- update the documentation:
- [x] all supported keywords and features are documented
- [x] there is no documentation for not yet implemented features
- [x] add section that discusses the differences between fix pimd/nvt and fix pimd/langevin
- [x] add a few small examples demonstrating the use of the command and some options
- remove all memory leaks:
- [x] deallocate all allocated arrays in destructor
- [x] delete all created classes in destructor
- [x] delete all added communicators in destructor
- [x] delete allocated storage before (re)allocating (requires initialization of pointers in with nullptr in constructor)
- [x] there must be no uninitialized memory accesses (check with valgrind, same for memory leaks)

### CODE REVIEW

**Comment by Yi-FanLi** [comment]
Hi @akohlmey and @athomps , I have renamed the fix style pimd to pimd/langevin based on your suggestions.

File: src/REPLICA/fix_pimd_langevin.h [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+  void setup(int) override;
+//   void post_neighbor() override;
+  void post_force(int) override;
+  void initial_integrate(int) override;
+  void final_integrate() override;
+  void end_of_step() override;
+
+  double compute_vector(int) override;
+
+  /* System setting variables */
```
Comment (akohlmey) [inline src/REPLICA/fix_pimd_langevin.h:44]: Why are the properties and functions below public and not protected?

Comment (Yi-FanLi) [reply src/REPLICA/fix_pimd_langevin.h:44]: The reason is that I learned the manner from fix_pimd.h when I started this work in late 2020. I have already pushed a commit that fixes this issue.

**Review by sjplimp — APPROVED** [review]

**Comment by sjplimp** [comment]
I approved this - thanks for submitting and separating it into two commands.   But I defer to @akohlmey on the implementation details.

**Comment by Yi-FanLi** [comment]
@akohlmey Hi Axel, could you please take a look at your convenience? Thanks a lot!

File: src/REPLICA/fix_pimd_langevin.cpp [hunk 2]
```diff
@@ -0,0 +1,1497 @@
...
+        ensemble = NPH;
+        tstat_flag = 0;
+        pstat_flag = 1;
+      } else if (strcmp(arg[i + 1], "npt") == 0) {
+        ensemble = NPT;
+        tstat_flag = 1;
+        pstat_flag = 1;
+      } else
+        error->universe_all(FLERR,
+                            "Unknown ensemble parameter for fix pimd/langevin. Only nve and nvt "
```
Comment (akohlmey) [inline src/REPLICA/fix_pimd_langevin.cpp]: Apparently "nph" and "npt" are supported now, too. So the error message should be updated.

File: src/REPLICA/fix_pimd_langevin.cpp [hunk 3]
```diff
@@ -0,0 +1,1497 @@
...
+    size_vector = 10;
+  else if (pstat_flag) {
+    if (pstyle == ISO) {
+      size_vector = 15;
+    } else if (pstyle == ANISO) {
+      size_vector = 17;
+    }
+  }
+  extvector = 1;
+  kBT = force->boltz * temp;
```
Comment (akohlmey) [inline src/REPLICA/fix_pimd_langevin.cpp]: I've looked through other parts of LAMMPS and the most common variable name for (fix nh, fix rigid) storing `k_B * T` is `kt`. In general, we try to reserve mixed capitalization for classes (but using "CamelCase" conventions), lower case for functions and variables and all caps for constants.

File: src/REPLICA/fix_pimd_langevin.cpp [hunk 4]
```diff
@@ -0,0 +1,1497 @@
...
+    error->universe_all(FLERR, "Unknown integrator parameter for fix pimd/langevin");
+  }
+
+  comm_init();
+
+  mass = new double[atom->ntypes + 1];
+
+  nmpimd_init();
+
+  Langevin_init();
```
Comment (akohlmey) [inline src/REPLICA/fix_pimd_langevin.cpp]: For consistency, please rename this function to all lowercase (`langevin_init()`). See my other comment.

File: doc/src/fix_pimd.rst [hunk 5]
```diff
@@ -243,3 +441,22 @@ Path Integrals, McGraw-Hill, New York (1965).
 
 **(Herman)** M. F. Herman, E. J. Bruskin, B. J. Berne, J Chem Phys, 76, 5150 (1982).
 
+.. _Bussi:
+
+**(Bussi)** G. Bussi, T. Zykova-Timan, M. Parrinello, J Chem Phys, 130, 074101 (2009).
+
+.. _Ceriotti2:
```
Comment (akohlmey) [inline doc/src/fix_pimd.rst]: References like this must be globally unique in Sphinx.  `Ceriotti2` is already used in `fix_gle.rst`, so please rename to `Ceriotti3`.

**Review by akohlmey — CHANGES_REQUESTED** [review]
There are only some minor, cosmetic changes. Changing kBT is optional. Otherwise, this looks ready to merge.

File: doc/src/fix_pimd.rst (hunk 5 above)
Comment (Yi-FanLi) [reply doc/src/fix_pimd.rst]: solved by [ee6b12e](https://github.com/lammps/lammps/pull/3660/commits/ee6b12ee300c8adb1c8cd62ba6f703b4a6bb73b1)

File: src/REPLICA/fix_pimd_langevin.cpp (hunk 4 above)
Comment (Yi-FanLi) [reply src/REPLICA/fix_pimd_langevin.cpp]: solved by 
a415d73

File: src/REPLICA/fix_pimd_langevin.cpp (hunk 3 above)
Comment (Yi-FanLi) [reply src/REPLICA/fix_pimd_langevin.cpp]: solved by [5ff6fd9](https://github.com/lammps/lammps/pull/3660/commits/5ff6fd9aad41084521a1a48381c6ebba6f584426)

File: src/REPLICA/fix_pimd_langevin.cpp (hunk 2 above)
Comment (Yi-FanLi) [reply src/REPLICA/fix_pimd_langevin.cpp]: solved by [f838677](https://github.com/lammps/lammps/pull/3660/commits/f83867787c6eb7ed8b509a54be25b3ee46e7290d)

**Review by akohlmey — APPROVED** [review]

**Pull request merged**